Jetpack Checklist: Implement Jetpack Backups task

tyxla opened the pull request:
#### Changes proposed in this Pull Request

* Implements the Jetpack Backups task for all use cases.

#### Previews

Note: primary buttons in the guided tour steps are no longer primary, and there might be small textual changes as feedback is addressed in the PR.

Pre-Rewind sites install and configure backups (VaultPress) automatically:
![](https://cldup.com/sINyGmEX3D.png)

Rewind sites require credentials, so there's a manual task for that:
![](https://cldup.com/Q63POyn9MH.png)

Clicking the "Do it" button leads to the Security Settings page:
![](https://cldup.com/9I5pupUdvx.png)

Clicking the "Add site credentials" button displays a confirmation section if the credentials have to be input manually:
![](https://cldup.com/kX2jQhCzuz.png)

Clicking the "Add site credentials" button displays a confirmation section if the credentials can be auto-provisioned (for example in Pressable):
![](https://cldup.com/WUt2goQ1EC.png)

After inputting credentials or auto-provisioning them successfuly, user gets a success message and a nudge to go back:
![](https://cldup.com/_Nz2NYAgrg.png)

Users can use the checklist to change credentials if they want:
![](https://cldup.com/6GdMysRgZZ.png)

Users can't do anything with the VP task, it's passive:
![](https://cldup.com/DQWd-hPdzb.png)

#### Testing instructions

* Checkout this branch or spin it up on calypso.live.
* Start a new Jurassic Ninja site with the latest bleeding edge and connect it to WP.com, pick the free plan.
* Open `/plans/:site` where `:site` is the slug of your site.
* Buy a Premium or Professional plan.
* Wait for the setup to finish, and verify that the Backups task is automatically setup - first displayed as "in progress", and when finished, displayed as "finished".
* Deactivate VaultPress from wp-admin.
* Go to VP MC and enable Rewind for that site.
* Go to `/plans/my-plan/:site` where `:site` is the slug of the site.
* Verify the "Backups & Scanning" task is now not done, and has a "Do it" button.
* Click the button.
* Verify the guided tour leads you to add credentials, as shown on the screenshots (for manual credentials entry).
* Verify the after you add credentials, and you click the button to go back to the checklist, the Backups and Scanning task is now displayed as complete.
* Pick a Pressable site.
* Go to `/settings/security/:site` where `:site` is the slug of the Pressable site.
* If you already have credentials for backups, click the green icon to edit them, and click "Revoke credentials" to remove them.
* Go to `/plans/my-plan/:site` where `:site` is the slug of the site.
* Verify the "Backups & Scanning" task is now not done, and has a "Do it" button.
* Click the button.
* Verify the guided tour leads you to add credentials, as shown on the screenshots (for auto provisioning).
* Verify the after credentials are auto-provisioned, and you click the button to go back to the checklist, the Backups and Scanning task is now displayed as complete.

tyxla committed:
Guided Tours: Introduce a JetpackBackupsRewindTour

tyxla committed:
Guided Tours: Register the new tour

tyxla committed:
Jetpack Checklist: Dynamic VP and Rewind task

tyxla committed:
Guided Tours: Polish JetpackBackupsRewindTour

tyxla committed:
Guided Tours: Reuse buttons in JetpackBackupsRewindTour

tyxla committed:
Guided Tours: Remove extra animations in JetpackBackupsRewindTour

tyxla committed:
JetpackBackupsRewindTour: Attempt to wait for credentials

jeffgolenski commented:
@shaunandrews @fditrapani I'd love to get some feedback and thoughts on how we can better create a "to do" hierarchy for our customers with the checklist here. For Jetpack, some items may be "blockers" that need to be completed for customers to get 100% value out of their plans. 

As noted in the video walkthrough I pinged you both on in slack, In order for Jetpack Backups and Restores to start working, we need to connect jetpack to each customer's respective server. The images @tyxla shared above, basically outline the flow right now. There's a checklist item for it, but it's not prominent and doesn't communicate to customers it needs to happen.

One thing I want to add is a more visual "urgent" list item to the security checklist that places it on a more important level than the rest of the items. Basically customers need to set this up or the backups and restores they're paying us for won't work. 

I want your feedback on adding a a pattern like this and if it makes sense to do this. In the future I'd love to entertain the idea of working this into onboarding possibily, but for now, I want to make sure people can visually see it needs to happen. 

![jp-checklist-scanning-alert](https://user-images.githubusercontent.com/214813/55915265-4833ae00-5bb7-11e9-995b-b2f559176b79.jpg)

(Marin, don't implement this just yet. Just want some feedback on this first. Thanks! )

cc @seear

tyxla committed:
ConnectedContinueToLastStep: Add fresh credentials state

fditrapani commented:
Thanks for the ping @jeffgolenski. There is some work happening on the navigator that can help you (p9Jlb4-Nh-p2). The work doesn't look 100% solidified but I think it's heading in a good direction that we'll want to impement on our checklist. 

![image](https://user-images.githubusercontent.com/6981253/55955792-596cd100-5c30-11e9-8faa-05f22169b304.png)

The thing to note is more emphasis on the task you're working on and less emphasis on the upcoming tasks. The current checklist on WP.com doesn't quite emphasize the current task but we did remove lots of the noise around the upcoming tasks. 

![image](https://user-images.githubusercontent.com/6981253/55955931-be282b80-5c30-11e9-845e-c91489ec31e4.png)

Something else to note about the screenshot above is there are some screens with multiple primary buttons. We need to make sure that pink is applied to the most important action. In a couple of instances it brings urgency to returning to the checklist vs completing the action.

![image](https://user-images.githubusercontent.com/6981253/55956061-0ba49880-5c31-11e9-8bb2-52c265bcf372.png)

tyxla committed:
JetpackBackupsRewindTour: Make buttons non-primary

tyxla commented:
Thanks @fditrapani - I've addressed your feedback on leaving a single primary button, none of the steps in this tour will have primary buttons, as we want to remove the main focus from getting back from the checklist:

![](https://cldup.com/1wgUc525db.png)

fditrapani commented:
Awesome. Thanks @tyxla.

tyxla commented:
This is ready for code and design review.

simison commented on client/my-sites/plans/current-plan/jetpack-checklist/constants.js:
```diff
@@ -58,3 +48,19 @@ export const JETPACK_CHECKLIST_TASKS = {
 };
+
+export const JETPACK_CHECKLIST_TASK_BACKUPS_REWIND = {
+	title: translate( 'Backups & Scanning' ),
```
Should this be "Backup and Scan" as per p1HpG7-6z0-p2 ?

jeffgolenski replied:
+1 per our new naming convention!

tyxla replied:
👍 Renamed.

simison commented on client/layout/guided-tours/style.scss:
```diff
@@ -124,11 +124,16 @@
 	}
-	.guided-tours__quit-button {
+	.guided-tours__quit-button,
+	a.button {
```
I know this wasn't introduced in this PR and suppose it's consistent with other tour-popups but feels odd we have custom styles like this outside the regular button style semantics & component.

<img width="298" alt="image" src="https://user-images.githubusercontent.com/87168/56028742-7503d480-5d21-11e9-8de3-b194807e8262.png">

I was expecting to see "text button": https://murieldesignsystem.blog/components/buttons/#types-and-anatomy

jeffgolenski replied:
I agree with @simison - if we can stay in line with our existing design patterns unless its absolutely necessary to deviate. Good eye!

tyxla replied:
I agree with you folks, but what we're doing in this PR is just reusing what is already there. The only change I'm doing here is actually stying that button, because before that it was appearing with white background and white text color. I prefer not changing anything that touches the existing guided tours in this PR. Does that make sense to you @simison and @jeffgolenski?

tyxla replied:
If we decide to change this, it will make sense to audit all existing guided tours. Let's leave that for another PR, cc @Automattic/onboarding-dev and @Automattic/onboarding-design

sirreal reviewed (commented):
This seems to work well. Two notes:

I think this is a pre-existing Guided Tours bug, but the point is in the wrong place:

![bad pointer](https://user-images.githubusercontent.com/841763/56030595-82b75b00-5d1d-11e9-9c8a-d2ca9847921b.png)

The first time I completed the flow and returned (by clicking the guided tour complete), the task did not reflect completion:

![on-return](https://user-images.githubusercontent.com/841763/56030596-82b75b00-5d1d-11e9-92a1-36bb19a44e6c.png)

It did on the next time I went through the flow.

fditrapani commented:
> I think this is a pre-existing Guided Tours bug, but the point is in the wrong place:

That's correct @sirreal. The guides can have some weird behaviours depending on the screen size.

seear commented:
<img width="481" alt="Screenshot 2019-04-12 at 13 18 00" src="https://user-images.githubusercontent.com/7767559/56036742-851db300-5d25-11e9-9db5-bb84f04ecf2d.png">

Styling the _Do it!_ button as _primary_ on the top item as in @jeffgolenski's comment is a great suggestion.

seear commented:
I gave this a test run. It is working great from the point of view of the Jetback Backup rollout. Some points:
* Anything we can do to highlight the importance of the credentials adding step, as per [Jeff's comment ](https://github.com/Automattic/wp-calypso/pull/32152#issuecomment-481877643) would be welcome.
* Also, anything we could do to reduce clicks between this checklist and the actual credentials adding screen would be good. After clicking _Do it!_, there is another click on the appropriate item in _Settings_ and yet another click for some terms-of-service.
* I saw a similar problem to Jon, where the checklist item didn't get marked as done until I navigated away from the checklist and back again.

Great work, thanks @tyxla!

tyxla committed:
Jetpack Checklist: Rename backups task

tyxla committed:
Jetpack Checklist: Rephrase rewind to restore

tyxla committed:
Jetpack Checklist: Rename completed backups task name

tyxla committed:
Jetpack Checklist: Rename VP task names

tyxla commented on client/my-sites/plans/current-plan/jetpack-checklist/index.js:
```diff
@@ -96,11 +126,7 @@ class JetpackChecklist extends PureComponent {
 
 						if ( ! task ) {
-							// UI does not support this task yet.
-							recordTracksEvent( 'calypso_jetpack_checklist_unsupported_task', {
```
Good question - I added it previously, but I think it doesn't make much sense to track those that way, so I've decided to remove it. It feels like and odd usage of tracks, and we should do a better job of keeping tasks on the backend and the client in sync.

tyxla commented:
> I think this is a pre-existing Guided Tours bug, but the point is in the wrong place:

Yeah, that can be encountered in some screen sizes indeed. As you and @fditrapani already confirmed, it's a pre-existing known bug with tours. Would be nice to get it fixed, but it's beyond this PR, I guess.

> The first time I completed the flow and returned (by clicking the guided tour complete), the task did not reflect completion:

Yeah, I'm getting reports for that for other tasks too, including from @seear below. Opened an issue to track that, as it's unrelated with this PR: #32248

tyxla commented:
> Anything we can do to highlight the importance of the credentials adding step, as per Jeff's comment would be welcome.

I'm happy to follow up in another PR with this. As @jeffgolenski requested, I haven't proceeded here, as it's not yet decided whether this will be the way to highlight important tasks.

> Also, anything we could do to reduce clicks between this checklist and the actual credentials adding screen would be good. After clicking Do it!, there is another click on the appropriate item in Settings and yet another click for some terms-of-service.

cc @jeffgolenski about this 😉 this will need some thinking and mockups.

> I saw a similar problem to Jon, where the checklist item didn't get marked as done until I navigated away from the checklist and back again.

Good point, opened an issue for that: #32248

tyxla commented:
Seems to me that feedback has either been addressed, or moved to other issues for cases where it's unrelated to the PR. 

Could I please get another code review and ✅ @simison @sirreal @seear? Thanks!

jeffgolenski commented:
Finally avoided jetpack manage problems and was able to test. Works wonderfully. I even downgraded my plan to free and re-upgraded my plan to professional and the checklist kept up. Nice! 

As mentioned above. design mods coming soon

jeffgolenski reviewed (approved):
Design review approved.

simison reviewed (commented):
>Wait for the setup to finish, and verify that the Backups task is automatically setup - first displayed as "in progress", and when finished, displayed as "finished".

This indeed happened. :thumbsup: 

<img width="568" alt="Screenshot 2019-04-12 at 17 57 08" src="https://user-images.githubusercontent.com/87168/56046914-ffadf900-5d4c-11e9-98f6-d9f8b5f2514e.png">

<img width="552" alt="Screenshot 2019-04-12 at 17 57 12" src="https://user-images.githubusercontent.com/87168/56046913-ffadf900-5d4c-11e9-9350-8c34515a689a.png">

Then I did the monitor step and got back to checklist, and found that the checklist item was again "loading" and monitor was unfinished:

<img width="438" alt="Screenshot 2019-04-12 at 17 57 30" src="https://user-images.githubusercontent.com/87168/56046932-076d9d80-5d4d-11e9-824e-f4a79b683347.png">

<img width="1096" alt="Screenshot 2019-04-12 at 17 58 14" src="https://user-images.githubusercontent.com/87168/56046951-17857d00-5d4d-11e9-8690-3faf1e85b5fe.png">

tyxla commented:
@simison I think that's the same one that Jon and Steve encountered - I've opened #32248 for it and will be tackling it separately.

simison commented:
Is it also expected that if I refresh this page, I get disconnection problem and spinners just keep spinning:

<img width="1479" alt="image" src="https://user-images.githubusercontent.com/87168/56047230-a5f9fe80-5d4d-11e9-86c9-5dcd8402af45.png">

Site's wp-admin loads fine via URL.

tyxla commented:
> Is it also expected that if I refresh this page, I get disconnection problem and spinners just keep spinning:

@jeffgolenski reported the same thing. It seems like an unrelated connection problem that I haven't been able to diagnose, but I asked in the Crew channel. FWIW, reconnecting fixes it, so it's unrelated with this PR.

simison reviewed (commented):
Re-connected the site, activated rewind..

Checklist doesn't infinite-spin anymore and I got to press "do it" for backups and scan.

I end up here:

<img width="1564" alt="Screenshot 2019-04-12 at 18 13 58" src="https://user-images.githubusercontent.com/87168/56047809-e9a13800-5d4e-11e9-8791-1919dd18515e.png">

Where I'm pretty confused what to press obviously. 😅 I choose to press big plus although other things on the page look much more like buttons (even primary pink ones) than the plus icon.

I then get this:

<img width="1627" alt="Screenshot 2019-04-12 at 18 14 21" src="https://user-images.githubusercontent.com/87168/56047873-0dfd1480-5d4f-11e9-819e-91fa4682ad73.png">

I now have 4 buttons to choose from 😵 I press "ok" because I know that'll bring me _onwards_, all other 3 buttons seem like _backwards_ (maybe?).

I fill in credentials and get a lot of green ✅ icons scattered across my screen:

<img width="1526" alt="Screenshot 2019-04-12 at 18 18 17" src="https://user-images.githubusercontent.com/87168/56048081-6c29f780-5d4f-11e9-816f-66185f52b568.png">

I press "let's do it" and get one final ✅ 

<img width="1090" alt="image" src="https://user-images.githubusercontent.com/87168/56048158-8fed3d80-5d4f-11e9-858a-f5087439cf8d.png">


...and with that, Marin, I give one for you too: ✅ 

Ship it! 😄

simison reviewed (approved):
Code looks good!

Testing experience otherwise was a bit rough but I'd expect visual issues raised up here be ironed out in follow ups. :-)

tyxla commented:
Thank you for the thorough testing and review 🙌 Will ship on Monday and continue iterating on making it less rough 😅

tyxla commented:
And I've opened #32249 to start thinking about how we can simplify and enhance the backups task to have less buttons and clicks and make it generally less rough 😅

simison commented:
Also noting that once I cancel monitor and remove SSH credentials, checklist is back to this again as expected:

<img width="1081" alt="image" src="https://user-images.githubusercontent.com/87168/56048977-35ed7780-5d51-11e9-8c9b-a810eed51838.png">


✅ ✅ ✅ 👍 🌮

sgomes commented:
@tyxla : I just saw this PR when looking at the commit log and wanted to point out that you didn't add the new tour to `layout/guided-tours/config.js`. This means that it'll be ignored by the guided tours selector (`client/state/ui/guided-tours/selectors`) and may not work properly if e.g. it ever gets a `when` condition.

tyxla commented:
Thanks for that @sgomes 👍 I'll make sure to add it soon.

tyxla commented:
#32274 registers the new guided tour properly, thanks again @sgomes!
